### PR TITLE
Add an example Smoke CO Alarm configuration to chef devices

### PR DESCRIPTION
**Pull request by ericzijian1994**
* Adds the Smoke CO Alarm device type to the list of matter device type.
* Adds an example Smoke CO Alarm device type configuration in chef/devices.

### CODE REVIEW

File: scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/CHIPClusters.h [hunk 1]
```diff
@@ -28,13 +28,17 @@
 namespace chip {
 namespace Controller {
 
+class DLL_EXPORT OnOffCluster : public ClusterBase
```
Comment (andy31415) [inline scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/CHIPClusters.h:31]: are these changes related?

File: scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/access.h [hunk 2]
```diff
@@ -49,6 +49,8 @@
     0x0000003E, /* Cluster: Operational Credentials, Attribute: NOCs, Privilege: administer */ \
     /* Cluster: Group Key Management, Attribute: GroupKeyMap, Privilege: view */ \
     /* Cluster: User Label, Attribute: LabelList, Privilege: view */ \
+    0x00000046, /* Cluster: ICD Management, Attribute: RegisteredClients, Privilege: administer */ \
```
Comment (andy31415) [inline scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/access.h:52]: why these changes?

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter [hunk 3]
```diff
@@ -0,0 +1,1666 @@
...
+    callback attribute bootReason;
+    callback attribute activeHardwareFaults;
+    callback attribute activeRadioFaults;
+    callback attribute activeNetworkFaults;
+    callback attribute testEventTriggersEnabled default = false;
+    ram      attribute featureMap default = 0;
+    ram      attribute clusterRevision default = 0x0001;
+  }
+
+  server cluster Switch {
```
Comment (andy31415) [inline examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter:1527]: I think this uses an old template - switch should not be part of endpoint 0

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter [hunk 4]
```diff
@@ -0,0 +1,1666 @@
...
+  readonly attribute bitmap32 featureMap = 65532;
+  readonly attribute int16u clusterRevision = 65533;
+
+  command SelfTestRequest(): DefaultSuccess = 0;
+}
+
+endpoint 0 {
+  device type rootdevice = 22, version 1;
+
+  server cluster Groups {
```
Comment (bzbarsky-apple) [inline examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter:1429]: Why is there Groups on endpoint 0?

Please go through this whole configuration carefully and make sure it makes sense.  This looks like a copy/paste from a broken config of some sort... and I know we recently fixed a bunch of such broken configs.

**Review by bzbarsky-apple — DISMISSED** [review]

File: scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/access.h (hunk 2 above)
Comment (ericzijian1994) [reply scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/access.h]: I was using the wrong Zap configuration, fixed.

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter (hunk 3 above)
Comment (ericzijian1994) [reply examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter]: Fixed.

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter (hunk 4 above)
Comment (ericzijian1994) [reply examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter]: Sorry, I missed the change, fixed.

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter [hunk 5]
```diff
@@ -0,0 +1,1590 @@
...
+    ram      attribute nameSupport;
+    callback attribute generatedCommandList;
+    callback attribute acceptedCommandList;
+    callback attribute attributeList;
+    ram      attribute featureMap default = 0;
+    ram      attribute clusterRevision default = 3;
+  }
+
+  server cluster Scenes {
```
Comment (bzbarsky-apple) [inline examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter:1518]: Smoke Alarm device should not have Scenes.

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter [hunk 6]
```diff
@@ -0,0 +1,1590 @@
...
+    callback attribute clientList default = 0;
+    callback attribute partsList default = 0;
+    callback attribute generatedCommandList default = 0;
+    callback attribute acceptedCommandList default = 0;
+    callback attribute attributeList default = 0;
+    ram      attribute featureMap default = 0;
+    ram      attribute clusterRevision default = 1;
+  }
+
+  server cluster PowerSource {
```
Comment (bzbarsky-apple) [inline examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter:1545]: Then this should list the Power Source device type too, right?

File: scripts/tools/zap/tests/inputs/all-clusters-app.zap
Comment (bzbarsky-apple) [inline scripts/tools/zap/tests/inputs/all-clusters-app.zap:1]: Just wondering why this test file had Smoke CO Alarm added to it... it's not necessarily a problem, but just making sure that the intent was not to add to some file that's actually used for something other than testing ZAP codegen.

Comment (ericzijian1994) [reply scripts/tools/zap/tests/inputs/all-clusters-app.zap:1]: Because the EMBER_AF_PLUGIN_SMOKE_CO_ALARM_SERVER macro is needed in stubs.cpp, just like the door lock.

Comment (bzbarsky-apple) [reply scripts/tools/zap/tests/inputs/all-clusters-app.zap:1]: What does stubs.cpp have to do with the _test_ .zap?  The test .zap should only be being used for testing ZAP generation.  Is someone (incorrectly) using its output for actual code?

File: scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/CHIPClusters.h (hunk 1 above)
Comment (ericzijian1994) [reply scripts/tools/zap/tests/outputs/all-clusters-app/app-templates/CHIPClusters.h]: Removed.

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter (hunk 5 above)
Comment (ericzijian1994) [reply examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter]: Removed.

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter (hunk 6 above)
Comment (ericzijian1994) [reply examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter:1402]: Can you explain this in more detail? Do the changes I'm making now fit?

File: scripts/tools/zap/tests/inputs/all-clusters-app.zap
Comment (ericzijian1994) [reply scripts/tools/zap/tests/inputs/all-clusters-app.zap:1]: It's removed.

File: examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter (hunk 6 above)
Comment (andy31415) [reply examples/chef/devices/rootnode_smokecoalarm_686fe0dcb8.matter:1406]: I believe this is to support multiple device types: the device type in smoecoalarm should also include "power source".

Can be updated as a followup.
Created https://github.com/project-chip/connectedhomeip/issues/28782
